src/executor/container_spawner.rs

```diff
@@ -0,0 +1,156 @@
+use bollard::{
+    container::{
+        Config, CreateContainerOptions, RemoveContainerOptions, StartContainerOptions,
+    },
+    service::ContainerInspectResponse,
+    Docker,
+};
+use std::{
+    default::Default,
+    error::Error,
+    future::Future,
+    pin::Pin,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+use tokio;
+use tokio::sync::oneshot;
+
+use crate::executor::{Executor, ExecutorSpawner};
+
+pub struct ContainerSpawner {
+    docker: Docker,
+    worker_counter: AtomicUsize,
+    image_name: String,
+    container_name: String,
+}
+
+impl ContainerSpawner {
+    pub fn new(image_name: String, container_name: String) -> Self {
+        ContainerSpawner {
+            docker: Docker::connect_with_local_defaults().unwrap(),
+            worker_counter: AtomicUsize::new(0),
+            image_name,
+            container_name,
+        }
+    }
+
+    // Create a Docker instance connected to the local Docker daemon.
+    pub async fn create_container(
+        docker: Docker,
+        image_name: String,
+        container_name: String,
+        id: usize,
+    ) -> Result<ContainerInspectResponse, Box<dyn Error>> {
+        let container_name = format!("{}_{}", container_name, id);
+
+        let config = Config {
+            image: Some(image_name),
+            ..Default::default()
+        };
+
+        // Create the container.
+        let create_container_options = CreateContainerOptions {
+            name: container_name.clone(),
+            platform: None,
+        };
+
+        docker
+            .create_container(Some(create_container_options), config.clone())
+            .await?;
+
+        docker
+            .start_container(
+                &container_name.clone(),
+                None::<StartContainerOptions<String>>,
+            )
+            .await?;
+
+        let container_info: ContainerInspectResponse =
+            docker.inspect_container(&container_name, None).await?;
+
+        Ok(container_info)
+    }
+}
+
+impl ExecutorSpawner for ContainerSpawner {
+    fn spawn_executor(&self) -> Pin<Box<dyn Future<Output = Executor> + Send>> {
+        // Using channel that onetime use, `oneshot`, to send container information
+        let (tx, rx) = oneshot::channel();
+
+        // These variables has to be cloned because it is moved into the async block
+        let docker_clone = self.docker.clone();
+        let image_name = self.image_name.clone();
+        let container_name = self.container_name.clone();
+        let id = self.worker_counter.fetch_add(1, Ordering::SeqCst);
+        tokio::spawn(async move {
+            if let Ok(container_info) =
+                ContainerSpawner::create_container(docker_clone, image_name, container_name, id)
+                    .await
+            {
+                let _ = tx.send(container_info);
```

Comment:
Where does that container_info go?

Reply:
on the line 96, It will be evaluated on the orchestrator.